### ISSUE TITLE
[bug] DoS Possibility if Client ID is known (Security)

### ISSUE THREAD

**Issue by miguelpduarte** [moscajs/aedes#597]
**System Information**
- Aedes: git pulled from master
- NodeJS: 14.15.5
- OS: Arch Linux
- Arch: x86_64

**Describe the bug**
Given that the `$SYS` topics are not read-only, an attacker can perform a Denial-of-Service attack on a client connected to `aedes`.

**To Reproduce**
1. Run the AEDES MQTT broker with default configs.
2. Listen to the `$SYS/aedes-cli/new/clients` topic to check for announcements of new clients. Let's consider the attacker "captures" the ID X of a client connecting to the broker
3. (as per the above step) Subscribe to a /test topic using an mqtt client so that we can see the ID being announced as a new client
4. Publish a message to `$SYS/aedes-cli-but-nope/new/clients` with a payload of the captured client ID X.

**Expected behavior**
The message published in `$SYS` is ignored since the topic is read-only for clients.

**Actual Behaviour**
The client with ID X is disconnected by the `aedes` broker. Relevant code:
https://github.com/moscajs/aedes/blob/c390099c27820cf81c97f4e9df9123cf9d558615/aedes.js#L150

Since the `serverId` will be different from the broker's own ID, the broker will close the connection.

In my tests using [mqtt](https://www.npmjs.com/package/mqtt) node library to subscribe to topics and publish messages, the client that was under attack would repeatedly try to reconnect. However, if we repeatedly publish the DoS messages with a small delay, the client is stuck always attempting to reconnect and will not receive messages to their subscribed topics. Example script:

```bash
while true; do sleep 0.5; mqtt_pub -h localhost -t '$SYS/aedes-cli-but-nope/new/clients' 'mqttjs_4b829ced'; done
```

In which `mqttjs_4b829ced` is the ID X we have "stolen" above.

**Additional context**
Screenshot that shows this happening:

![image](https://user-images.githubusercontent.com/5095627/111816399-b53c3c00-88d4-11eb-9c59-ef7d8f0e70ec.png)

A cleaned up snippet of an iteration of this attack:

```
------CLIENT INIT, client connected------
Client mqttjs_d8d76c44 CONNECTED, broker aedes-cli
Client BROKER_aedes-cli PUBLISH mqttjs_d8d76c44 on $SYS/aedes-cli/new/clients, broker aedes-cli
CONNACK sent to mqttjs_d8d76c44, broker aedes-cli
Client mqttjs_d8d76c44 READY, broker aedes-cli
Client mqttjs_e431dff2 DISCONNECTED, broker aedes-cli
Client BROKER_aedes-cli PUBLISH mqttjs_e431dff2 on $SYS/aedes-cli/disconnect/clients, broker aedes-cli
Client mqttjs_d8d76c44 PUBLISH mqttjs_e431dff2 on $SYS/aedes-cli-but-nope/new/clients, broker aedes-cli
Client mqttjs_e431dff2 UNSUBSCRIBED to: , broker aedes-cli
Client BROKER_aedes-cli PUBLISH {"clientId":"mqttjs_e431dff2","subs":["/test"]} on $SYS/aedes-cli/new/unsubscribes, broker aedes-cli
Client mqttjs_d8d76c44 DISCONNECTED, broker aedes-cli
Client BROKER_aedes-cli PUBLISH mqttjs_d8d76c44 on $SYS/aedes-cli/disconnect/clients, broker aedes-cli
------CLIENT INIT, client connected------
Client mqttjs_e431dff2 CONNECTED, broker aedes-cli
Client BROKER_aedes-cli PUBLISH mqttjs_e431dff2 on $SYS/aedes-cli/new/clients, broker aedes-cli
CONNACK sent to mqttjs_e431dff2, broker aedes-cli
Client mqttjs_e431dff2 READY, broker aedes-cli
Client mqttjs_e431dff2 SUBSCRIBED to: /test, broker aedes-cli
Client BROKER_aedes-cli PUBLISH {"clientId":"mqttjs_e431dff2","subs":[{"topic":"/test","qos":0}]} on $SYS/aedes-cli/new/subscribes, broker aedes-cli
```

For this run, `mqttjs_e431dff2` was the ID of the attacked client, and `mqttjs_d8d76c44` is the ID of the attacker.

We can see that the attacker connects and receives a CONNACK. It is considered READY, and meanwhile the attacked client is DISCONNECTED (probably out of order logging here) due to the PUBLISH done by the attacker (that appears in the following lines). The client later on reconnects just fine, but if this publishing is repeated (as mentioned above) the attacked client will be unable to receive the subscribed messages due to being almost constantly offline.

**Comment by gnought** [comment]
Security awareness is our priority. We are very appreciate your detailed report and share the insight. Although it is a little bit huge, we will try our effort and look into it. By the way it is very helpful if anyone proposes a solution and draft/post a fix.

**Comment by robertsLando** [comment]
@gnought @mcollina does someone know how other brokers like mosquitto handles this?

**Comment by gnought** [comment]
I think the main problem here is to allow clients publish $SYS topic message.
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349844

**Comment by mcollina** [comment]
The $SYS topic space was never designed to be read-only in Aedes.

There is an authorizePublish utility to add all level of ACL needed to secure an Aedes deployment, including the $SYS topic. Nevertheless, adding this protection to the default implementation of authorizePublish code would be nice.

---

Sending potential security issue in public is not good practice as it exposes everybody without giving the time for the project to fix it. Be more careful next time.

@gnought we should really add a SECURITY.md file to allow people to send these via mail.

**Comment by miguelpduarte** [comment]
@mcollina I'm sorry, the issue can be deleted if you prefer this to be discussed privately. I believe [admins can do this](https://docs.github.com/en/github/managing-your-work-on-github/deleting-an-issue). Otherwise I can edit the issue description and delete the previous revision.

**Comment by robertsLando** [comment]
Based on the docs linked by @gnought :

Non normative comment
·         $SYS/ has been widely adopted as a prefix to topics that contain Server-specific information or control APIs
·         Applications cannot use a topic with a leading $ character for their own purposes
 
Non normative comment
·         A subscription to “#” will not receive any messages published to a topic beginning with a $
·         A subscription to “+/monitor/Clients” will not receive any messages published to “$SYS/monitor/Clients”
·         A subscription to “$SYS/#” will receive messages published to topics beginning with “$SYS/”
·         A subscription to “$SYS/monitor/+” will receive messages published to “$SYS/monitor/Clients”
·         For a Client to receive messages from topics that begin with $SYS/ and from topics that don’t begin with a $, it has to subscribe to both “#” and “$SYS/#”

**Comment by mcollina** [comment]
> @mcollina I'm sorry, the issue can be deleted if you prefer this to be discussed privately. I believe [admins can do this](https://docs.github.com/en/github/managing-your-work-on-github/deleting-an-issue). Otherwise I can edit the issue description and delete the previous revision.

No worries. Despite your claims, this is not a security vulnerability.

**Comment by robertsLando** [comment]
Just submitted a fix for this #622
